File: src/main/java/com/picimako/devkitplus/reference/CallMatcherReferenceContributor.java
//Copyright 2021 Tamás Balog. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.

package com.picimako.devkitplus.reference;

import static com.intellij.json.psi.JsonPsiUtil.stripQuotes;
import static com.intellij.psi.util.PsiLiteralUtil.isUnsafeLiteral;
import static com.picimako.devkitplus.CallMatcherUtil.ARGUMENT_OF_CALL_MATCHER_PATTERN;
import static com.picimako.devkitplus.CallMatcherUtil.CALL_MATCHER_EXACT_INSTANCE_MATCHER;
import static com.picimako.devkitplus.CallMatcherUtil.CALL_MATCHER_STATIC_MATCHER;
import static com.picimako.devkitplus.CallMatcherUtil.filterByNonStatic;
import static com.picimako.devkitplus.CallMatcherUtil.filterByStatic;

import java.util.Arrays;
import java.util.Optional;
import java.util.function.Supplier;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.JavaPsiFacade;
import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiElementResolveResult;
import com.intellij.psi.PsiExpression;
import com.intellij.psi.PsiLiteralExpression;
import com.intellij.psi.PsiMethod;
import com.intellij.psi.PsiMethodCallExpression;
import com.intellij.psi.PsiPolyVariantReference;
import com.intellij.psi.PsiReference;
import com.intellij.psi.PsiReferenceBase;
import com.intellij.psi.PsiReferenceContributor;
import com.intellij.psi.PsiReferenceProvider;
import com.intellij.psi.PsiReferenceRegistrar;
import com.intellij.psi.ResolveResult;
import com.intellij.psi.search.ProjectScope;
import com.intellij.psi.util.PsiTreeUtil;
import com.intellij.util.ProcessingContext;
import com.intellij.util.SmartList;
import com.siyeh.ig.callMatcher.CallMatcher;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Adds references to the arguments of {@link CallMatcher} static factory methods: {@code staticCall}, {@code instanceCall}, {@code exactInstanceCall}.
 * The first one is the fully qualified name of the referenced class, while the rest are methods of said class.
 * <p>
 * Upon resolving the references, the IDE navigates to the resolved class definition.
 * <p>
 * In case of methods, if the call is {@link CallMatcher#exactInstanceCall(String, String...)},
 * then it provides references for methods exactly in the referenced class,
 * otherwise references are provided for methods in the super classes too.
 * <p>
 * If the call is {@link CallMatcher#staticCall(String, String...)}, it provides references only for static methods.
 *
 * @since 0.1.0
 */
public class CallMatcherReferenceContributor extends PsiReferenceContributor {

    @Override
    public void registerReferenceProviders(@NotNull PsiReferenceRegistrar registrar) {
        registrar.registerReferenceProvider(
            ARGUMENT_OF_CALL_MATCHER_PATTERN,
            new PsiReferenceProvider() {
                @Override
                public PsiReference @NotNull [] getReferencesByElement(@NotNull PsiElement element, @NotNull ProcessingContext context) {
                    var parentCall = PsiTreeUtil.getParentOfType(element, PsiMethodCallExpression.class);
                    if (parentCall == null) return PsiReference.EMPTY_ARRAY;

                    var callMatcherArguments = parentCall.getArgumentList();
                    final var reference = new SmartList<PsiReference>();
                    //If the current literal is the first argument (the class FQN) of the CallMatcher call
                    if (element.getManager().areElementsEquivalent(element, callMatcherArguments.getExpressions()[0])) {
                        if (!isUnsafeLiteral((PsiLiteralExpression) element)) {
                            Optional.ofNullable(findClass(element))
                                .ifPresent(referencedClass -> reference.add(new CallMatcherArgReference(element, () -> new PsiElement[]{referencedClass})));
                        }
                    }
                    //If there is at least one method name specified
                    else if (callMatcherArguments.getExpressionCount() > 1 && !isUnsafeLiteral((PsiLiteralExpression) element)) {
                        var className = parentCall.getArgumentList().getExpressions()[0];
                        //If the classname is a String we can simplify find the class by it, otherwise first we have to evaluate the expression
                        var refClass = className instanceof PsiLiteralExpression
                            ? Optional.ofNullable(findClass(className))
                            : Optional.ofNullable(evaluate(className)).map(cls -> findClass(cls.toString(), element.getProject()));
                        refClass.ifPresent(referencedClass -> reference.add(new CallMatcherArgReference(element, () -> getMethodsByName(element, referencedClass, parentCall))));
                    }
                    return !reference.isEmpty() ? reference.toArray(PsiReference[]::new) : PsiReference.EMPTY_ARRAY;
                }
            });
    }

    private static Object evaluate(PsiExpression expression) {
        return JavaPsiFacade.getInstance(expression.getProject()).getConstantEvaluationHelper().computeConstantExpression(expression, true);
    }

    @Nullable
    public static PsiClass findClass(@NotNull PsiElement element) {
        return findClass(stripQuotes(element.getText()), element.getProject());
    }

    @Nullable
    private static PsiClass findClass(@NotNull String text, Project project) {
        return JavaPsiFacade.getInstance(project).findClass(text, ProjectScope.getAllScope(project));
    }

    @NotNull
    public static PsiMethod[] getMethodsByName(PsiElement element, @NotNull PsiClass referencedClass, PsiMethodCallExpression parentCall) {
        var methodsInClass = referencedClass.findMethodsByName(stripQuotes(element.getText()), !CALL_MATCHER_EXACT_INSTANCE_MATCHER.matches(parentCall));
        return CALL_MATCHER_STATIC_MATCHER.matches(parentCall) ? filterByStatic(methodsInClass) : filterByNonStatic(methodsInClass);
    }

    /**
     * Reference implementation to for class FQN and method name string literals in {@code CallMatcher} factory method arguments.
     */
    private static final class CallMatcherArgReference extends PsiReferenceBase<PsiElement> implements PsiPolyVariantReference {
        private final Supplier<PsiElement[]> elementsToResolveTo;

        public CallMatcherArgReference(@NotNull PsiElement element, Supplier<PsiElement[]> elementsToResolveTo) {
            super(element, TextRange.create(1, element.getTextRange().getLength() - 1), true);
            this.elementsToResolveTo = elementsToResolveTo;
        }

        @Override
        public ResolveResult @NotNull [] multiResolve(boolean incompleteCode) {
            return !incompleteCode
                ? Arrays.stream(elementsToResolveTo.get()).map(PsiElementResolveResult::new).toArray(ResolveResult[]::new)
                : ResolveResult.EMPTY_ARRAY;
        }

        @Override
        public @Nullable PsiElement resolve() {
            ResolveResult[] resolveResults = multiResolve(false);
            return resolveResults.length == 1 ? resolveResults[0].getElement() : null;
        }
    }
}

File: src/main/kotlin/com/picimako/justkitting/reference/CallMatcherReferenceContributor.kt
//Copyright 2023 Tamás Balog. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.picimako.justkitting.reference

import com.intellij.json.psi.JsonPsiUtil
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.TextRange
import com.intellij.psi.*
import com.intellij.psi.search.ProjectScope
import com.intellij.psi.util.PsiLiteralUtil.isUnsafeLiteral
import com.intellij.psi.util.PsiTreeUtil.getParentOfType
import com.intellij.util.ProcessingContext
import com.intellij.util.SmartList
import com.picimako.justkitting.CallMatcherUtil
import java.util.function.Supplier

/**
 * Adds references to the arguments of [com.siyeh.ig.callMatcher.CallMatcher] static factory methods: `staticCall`, `instanceCall`, `exactInstanceCall`.
 * The first one is the fully qualified name of the referenced class, while the rest are methods of said class.
 *
 * Upon resolving the references, the IDE navigates to the resolved class definition.
 *
 * In case of methods, if the call is [com.siyeh.ig.callMatcher.CallMatcher.exactInstanceCall],
 * then it provides references for methods exactly in the referenced class,
 * otherwise references are provided for methods in the super classes too.
 *
 * If the call is [com.siyeh.ig.callMatcher.CallMatcher.staticCall], it provides references only for static methods.
 *
 * @since 0.1.0
 */
class CallMatcherReferenceContributor : PsiReferenceContributor() {

    override fun registerReferenceProviders(registrar: PsiReferenceRegistrar) {
        registrar.registerReferenceProvider(
            CallMatcherUtil.ARGUMENT_OF_CALL_MATCHER_PATTERN,
            object : PsiReferenceProvider() {
                override fun getReferencesByElement(element: PsiElement, context: ProcessingContext): Array<PsiReference> {
                    val parentCall = getParentOfType(element, PsiMethodCallExpression::class.java)
                        ?: return PsiReference.EMPTY_ARRAY

                    val callMatcherArguments = parentCall.argumentList
                    val reference = SmartList<PsiReference>()

                    //If the current literal is the first argument (the class FQN) of the CallMatcher call
                    if (element.manager.areElementsEquivalent(element, callMatcherArguments.expressions[0])) {
                        if (!isUnsafeLiteral(element as PsiLiteralExpression)) {
                            findClass(element)?.let {
                                reference.add(CallMatcherArgReference(element) { arrayOf(it) })
                            }
                        }
                    } else if (callMatcherArguments.expressionCount > 1 && !isUnsafeLiteral(element as PsiLiteralExpression)) {
                        val className = parentCall.argumentList.expressions[0]
                        //If the classname is a String we can simply find the class by it, otherwise first we have to evaluate the expression
                        val referencedClass: PsiClass? =
                            if (className is PsiLiteralExpression) findClass(className)
                            else evaluate(className)?.let { findClass(it.toString(), element.getProject()) }

                        //Mapping the PsiMethods to 'it', so that they are passed as PsiElements
                        referencedClass?.let { reference.add(CallMatcherArgReference(element) { getMethodsByName(element, it, parentCall).map { it }.toTypedArray() }) }
                    }
                    return if (reference.isNotEmpty()) reference.toTypedArray() else PsiReference.EMPTY_ARRAY
                }
            })
    }

    /**
     * Reference implementation to for class FQN and method name string literals in `CallMatcher` factory method arguments.
     */
    private class CallMatcherArgReference(element: PsiElement, private val elementsToResolveTo: Supplier<Array<PsiElement>>)
        : PsiReferenceBase<PsiElement?>(element, TextRange.create(1, element.textRange.length - 1), true), PsiPolyVariantReference {

        override fun multiResolve(incompleteCode: Boolean): Array<ResolveResult> {
            return if (!incompleteCode)
                elementsToResolveTo.get()
                    .map { element: PsiElement -> PsiElementResolveResult(element) }
                    .toTypedArray()
            else ResolveResult.EMPTY_ARRAY
        }

        override fun resolve(): PsiElement? {
            val resolveResults = multiResolve(false)
            return if (resolveResults.size == 1) resolveResults[0].element else null
        }
    }

    companion object {
        @JvmStatic
        fun findClass(element: PsiElement): PsiClass? {
            return findClass(JsonPsiUtil.stripQuotes(element.text), element.project)
        }

        private fun evaluate(expression: PsiExpression): Any? {
            return JavaPsiFacade.getInstance(expression.project).constantEvaluationHelper.computeConstantExpression(expression, true)
        }

        private fun findClass(text: String, project: Project): PsiClass? {
            return JavaPsiFacade.getInstance(project).findClass(text, ProjectScope.getAllScope(project))
        }

        private fun getMethodsByName(element: PsiElement, referencedClass: PsiClass, parentCall: PsiMethodCallExpression?): Array<PsiMethod> {
            val methodsInClass = referencedClass.findMethodsByName(
                JsonPsiUtil.stripQuotes(element.text),
                !CallMatcherUtil.CALL_MATCHER_EXACT_INSTANCE_MATCHER.matches(parentCall))

            return if (CallMatcherUtil.CALL_MATCHER_STATIC_MATCHER.matches(parentCall))
                CallMatcherUtil.filterByStatic(methodsInClass)
            else
                CallMatcherUtil.filterByNonStatic(methodsInClass)
        }
    }
}
